Design/y_e_data_storage_modes.md

```diff
@@ -0,0 +1,151 @@
+
+- `Histogram` currently provides functions for obtaining variances as well as standard deviations, e.g., `countVariances()` and `countStandardDeviations()`.
+  If the internal `E` vector were converted to variances by external means (as it is done currently), these access methods would do the wrong thing.
+
+- `HistogramData` provides means for converting between standard deviations and variances, in particular types that do the conversion automatically.
+  These should be used by client code.
+
+- If summing spectra were done using `Histogram::operator+()`, propagating the uncertainties would involve a `sqrt` computation for every `+` operation.
+  In the example of `DiffractionFocussing2` this could yield a significant overhead, and it is not obvious how to solve this:
+
+  - The algorithm could do the summing itself, summing `Y` and variances separately.
+    This would cause no performance loss, and the code would be reasonably clean, e.g.,
+
+    ```cpp
+    for (size_t i=0; i<inputWS.getNumberHistograms(); ++i) {
+      counts += inputWS.counts(i);
+      variances += inputWS.countVariances(i);
+    }
+    ```
+
+  - We could provide optimized means for summing many spectra as part of the `HistogramData` module.
+    For example,
+
+    ```cpp
+    namespace HistogramData {
+    Histogram sumHistograms(const std::vector<Histogram> &histograms);
+    }
+    ```
+
+    This would be a clean solution without performance loss.
+    Unfortunately, there will be an implicit performance and ressource issue if the input is an `EventWorkspace`!
+    Obtaining a `Histogram` from the input workspace will create the `Y` and `E` data on the fly (this needs to be done in any solution).
+    However, to create the vector of histograms as input for `sumHistograms()` we need to **store all histograms at the same time**.
+    This will increase memory consumption and reduce cache locality of the computation.
+
+
+## Count and frequency data
+
+### Status
+
+- Data in `Workspace2D` be represent counts or frequencies.
+- Conversions between the two are done with the algorithms `ConvertToDistribution` and `ConvertFromDistribution`, or the helper function `WorkspaceHelpers::makeDistribution()`.
+- `MatrixWorkspace::isDistribution()` can be used for checking the state, and `MatrixWorkspace::setDistribution()` for setting the corresponding flag (note that this does not actually transform the data!).
+- There is a validator that verifies the state of a workspace, `RawCountValidator`, however is is used by relatively few algorithms.
+- A considerable number of algorithms use `MatrixWorkspace::isDistribution()` to adapt their behavior, in most cases this adaption is a simple conversion to the desired data storage type.
+- For a large number of algorithms there are no checks, and it is not documented whether or not the work only with counts, only with frequencies, or both.
+
```

Comment:
Well, that is not the only example, what is in my opinion much worse is `bool MatrixWorkspace::isCommonBins()` (don't look at it if you want to avoid some nightmares).